builtin/credential/aws/path_config_rotate_root.go

```diff
@@ -0,0 +1,223 @@
+		// could add the ability to specify the region either on the client config or as part of the
+		// inbound rotation call.
+		Region: aws.String(region),
+
+		// Prevents races.
+		HTTPClient: cleanhttp.DefaultClient(),
+	}
+	sess, err := session.NewSession(awsConfig)
+	if err != nil {
+		return nil, err
+	}
+	iamClient := getIAMClient(sess)
+
+	// Get the current user's name since it's required to create an access key.
+	// Empty input means get the current user.
+	var getUserInput iam.GetUserInput
+	getUserRes, err := iamClient.GetUser(&getUserInput)
+	if err != nil {
+		return nil, errwrap.Wrapf("error calling GetUser: {{err}}", err)
+	}
+	if getUserRes == nil {
+		return nil, fmt.Errorf("nil response from GetUser")
+	}
+	if getUserRes.User == nil {
+		return nil, fmt.Errorf("nil user returned from GetUser")
+	}
+	if getUserRes.User.UserName == nil {
+		return nil, fmt.Errorf("nil UserName returned from GetUser")
+	}
+
+	// Create the new access key and secret.
+	createAccessKeyInput := iam.CreateAccessKeyInput{
+		UserName: getUserRes.User.UserName,
+	}
+	createAccessKeyRes, err := iamClient.CreateAccessKey(&createAccessKeyInput)
+	if err != nil {
+		return nil, errwrap.Wrapf("error calling CreateAccessKey: {{err}}", err)
+	}
+	if createAccessKeyRes.AccessKey == nil {
+		return nil, fmt.Errorf("nil response from CreateAccessKey")
+	}
+	if createAccessKeyRes.AccessKey.AccessKeyId == nil || createAccessKeyRes.AccessKey.SecretAccessKey == nil {
+		return nil, fmt.Errorf("nil AccessKeyId or SecretAccessKey returned from CreateAccessKey")
+	}
+
+	// We're about to attempt to store the newly created key and secret, but just in case we can't,
+	// let's clean up after ourselves.
+	storedNewConf := false
+	var errs error
+	defer func() {
+		if storedNewConf {
+			return
+		}
+		// Attempt to delete the access key and secret we created but couldn't store and use.
+		deleteAccessKeyInput := iam.DeleteAccessKeyInput{
+			AccessKeyId: createAccessKeyRes.AccessKey.AccessKeyId,
+			UserName:    getUserRes.User.UserName,
+		}
+		if _, err := iamClient.DeleteAccessKey(&deleteAccessKeyInput); err != nil {
+			// Include this error in the errs returned by this method.
+			errs = multierror.Append(errs, fmt.Errorf("error deleting newly created but unstored access key ID %s: %s", *createAccessKeyRes.AccessKey.AccessKeyId, err))
+		}
+	}()
+
+	// Now get ready to update storage, doing everything beforehand so we can minimize how long
+	// we need to hold onto the lock.
+	newEntry, err := b.configClientToEntry(clientConf)
+	if err != nil {
+		errs = multierror.Append(errs, errwrap.Wrapf("error generating new client config JSON: {{err}}", err))
+		return nil, errs
+	}
+
+	oldAccessKey := clientConf.AccessKey
+	clientConf.AccessKey = *createAccessKeyRes.AccessKey.AccessKeyId
+	clientConf.SecretKey = *createAccessKeyRes.AccessKey.SecretAccessKey
+
+	// Now we will be mutating the store and caches, so we will need to prevent reads and mutations.
```

Comment:
+1 on this. I'd agree that we should simply hold a write lock for the entire duration since this is not an operation that would be performed that often.

Reply:
Done